api/resources_portal/middleware/oauth.py

```diff
@@ -0,0 +1,97 @@
+import os
+import urllib
+
+from django.conf import settings
+from django.contrib.auth import login
+
+import furl
+import orcid
+import requests
+
+from resources_portal.models.grant import Grant
+from resources_portal.models.organization import Organization
+from resources_portal.models.user import User
+
+CLIENT_ID = settings.CLIENT_ID
+CLIENT_SECRET = settings.CLIENT_SECRET
+
+
+def remove_code_parameter_from_uri(url):
+    url_obj = furl.furl(url)
+    url_obj.remove(["code"])
+    return urllib.parse.unquote(url_obj.url)
+
+
+class OAuthMiddleWare:
+
+    """ Intercepts requests containing a "code" parameter provided by ORCID OAuth.
+    The authorization code is checked against ORCID and will provide an access token."""
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        if "code" not in request.GET.keys():
+            response = self.get_response(request)
+            return response
+        else:
+            authorization_code = request.GET["code"]
+
+            # remove code param from uri so the redirect_uris will match
+            uri = remove_code_parameter_from_uri(request.build_absolute_uri())
+
+            data = {
+                "client_id": CLIENT_ID,
+                "client_secret": CLIENT_SECRET,
+                "grant_type": "authorization_code",
+                "code": authorization_code,
+                "redirect_uri": uri,
+            }
+
+            # get user orcid info
+            URL = "https://sandbox.orcid.org/oauth/token"
```

Comment:
We're probably also going to want to make this an env var so we can hit the not-sandbox URL in prod.

Reply:
Cool- done.